consumer_monitor.go

```diff
@@ -0,0 +1,280 @@
+
+			if m.scheduler.failureCount != 0 {
+				// Revert the scheduling change.
+				m.scheduler.failureCount = 0
+				m.scheduler.ticker.Reset(m.scheduler.standardInterval)
+			}
+		}
+	}
+}
+
+func (m *consumerMonitor) updateSubscriptions(event *subscriptionEvent) {
+	if event.subscribed {
+		m.subscriptions[event.topic] = append(m.subscriptions[event.topic], event.partition)
+		return
+	}
+
+	partitions, ok := m.subscriptions[event.topic]
+	if !ok {
+		return
+	}
+
+	for i, p := range partitions {
+		if p == event.partition {
+			partitions = append(partitions[:i], partitions[i+1:]...)
+			break
+		}
+	}
+
+	if len(partitions) == 0 {
+		delete(m.subscriptions, event.topic)
+		return
+	}
+
+	m.subscriptions[event.topic] = partitions
+}
+
+// Fetch and return the latest committed offsets by the consumer group
+// as well as those at the end of the log for the respective topics.
+func (m *consumerMonitor) fetchOffsets() (map[string][]offset, error) {
+	fetchedOffsets, err := m.consumerOffsetFetcher.ListConsumerGroupOffsets(m.consumerGroup, m.subscriptions)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching consumer group offsets from client: %w", err)
+	}
+
+	if fetchedOffsets.Err != sarama.ErrNoError {
+		return nil, fmt.Errorf("error fetching consumer group offsets from server: %w", fetchedOffsets.Err)
+	}
+
+	topicOffsets := map[string][]offset{}
+	for topic := range m.subscriptions {
+		partitions, ok := fetchedOffsets.Blocks[topic]
+		if !ok {
+			return nil, fmt.Errorf("requested consumer offsets for topic %q were not fetched", topic)
+		}
+
+		var offsets []offset
+		for partition, block := range partitions {
+			if block.Err != sarama.ErrNoError {
+				return nil, fmt.Errorf("error fetching consumer group offsets for partition %d of topic %q from server: %w",
+					partition, topic, block.Err)
+			}
+
+			topicOffset, err := m.topicOffsetFetcher.GetOffset(topic, partition, sarama.OffsetNewest)
+			if err != nil {
+				return nil, fmt.Errorf("error fetching topic offset for partition %d of topic %q: %w",
+					partition, topic, err)
+			}
+
+			offsets = append(offsets, offset{
+				Partition: partition,
+				Consumer:  block.Offset,
+				Topic:     topicOffset,
+			})
+		}
+
+		topicOffsets[topic] = offsets
+	}
+
+	return topicOffsets, nil
+}
+
+func (m *consumerMonitor) updateConsumerStatus(fetchedOffsets map[string][]offset) {
+	m.topicsLock.Lock()
+	defer m.topicsLock.Unlock()
+
+	for _, topic := range m.topics {
+		offsets, ok := fetchedOffsets[topic.Name]
+		if !ok {
+			// No active subscriptions for the given topic at the time.
+			topic.partitionLag = map[int32]int64{}
+			continue
+		}
+
+		for _, offset := range offsets {
+			// Only store the lag if it exceeds the configured threshold or is invalid.
+			// The next offset in the topic will always be greater than or equal to the last committed offset of the consumer.
+			lag := offset.Topic - offset.Consumer
+			if lag > topic.lagTolerance || lag < 0 {
```

Comment:
**question:** Why are we checking `lag < 0` if `.Topic` is always greater than `.Consumer` ?

Reply:
There's this very improbable scenario where something goes terribly wrong and for some reason the next topic offset (or the one returned as response of such requests) is lower than the latest committed consumer offset.

In general, we should be safe because we firstly fetch the consumer group offsets and then the topic ones (which would in turn allow for those to be even further shifted). However, if the reverse ever happens we'd never notice it unless we check if the lag is "negative" here and I believe it's reasonable to report a specific error in such cases.